Translation Integration with Transifex - Phase 1

mkangia opened the pull request:
This is the first phase of integration.

This adds the ability to send files to transifex for translation for a particular build of an app.
Heads up that the implementation for this might add things when follow up PR is done, for ex: add ability to send translation files with translations for target languages as well or add better occurrences for the strings.

A follow up PR would address pulling files from transifex and then adding those translations

mkangia committed:
Add transifex Client

mkangia committed:
Add ability to generate and push files to transifex

mkangia committed:
Add view to submit files for translation

mkangia committed:
add expectation for fields in the constant [skip ci]

mkangia committed:
lint [skip ci]

mkangia committed:
add toggle to restrict access

mkangia committed:
use lazy instead of noop

mkangia committed:
fix method called

mkangia committed:
fix view for getting saved versions of the app

mkangia committed:
replicate workflow at transifex and minor fixes

replicate workflow -
- allow choosing transifex project corresponding to source lang
- remove target lang

minor fixes-
- add version if not available

mkangia committed:
minor improvments

mkangia committed:
Ensure source language and transifex account details for project

mkangia committed:
handle difference in lang code

mkangia committed:
minor fixes

revert formplayer url changed
use ugettext_lazy instead of ugettext
add project_slug param
update example for new expected structure

mkangia commented on custom/icds/translations/integrations/client.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class TransifexApiClient():
+    def __init__(self, token, orgranization, project):
```
eye of the tiger, thanks 👍

mkangia commented on custom/icds/translations/integrations/client.py:
```diff
@@ -0,0 +1,43 @@
+
+    def upload_resource(self, path_to_pofile, resource_slug, resource_name):
+        url = "https://www.transifex.com/api/2/project/{}/resources".format(self.project)
+        content = open(path_to_pofile, 'r').read()
```
ya, i don't think it will retain keep any reference to the file.

> is the pofile kept as a temporary file on the filesystem?
No its not. Its generated in the current directory and then deleted at the end of the process here itself. I would like us to remove them here because they can be many in number and in case this rolls out as a feature for multiple domains and is used, it would result in quite some files which would stay around till they are removed automatically.
Or can also make them temp file but still remove them if that sounds better, though not sure why, possibly in case this process fails and does not reach the step to remove them manually?

mkangia commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -0,0 +1,200 @@
+Unique_ID = namedtuple('UniqueID', 'type id')
+
+
+class Transifex():
```
I think having `Transifex` still makes sense to handle what needs to be sent and then processing whats received whereas `TransifexClient` just handles sending/pulling from transifex.
Agree about extracting the functionality to generate po files. will do that.

mkangia committed:
review fixes

- underscore all private methods
- fix spell
- better var naming

mkangia committed:
review fixes 2

instead of manually finding build id just let it be a memoized property

mkangia commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -0,0 +1,200 @@
+        # get app or the app for the build
+        self.find_build_id()
+        from corehq.apps.app_manager.dbaccessors import get_app
+        app = get_app(self.domain, self.app_id_to_build)
```
haha.. ohkay!

mkangia commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -0,0 +1,200 @@
+        for header_row in expected_bulk_app_sheet_headers(app):
+            self.headers[header_row[0]] = header_row[1]
+        self.set_sheet_name_to_module_or_form_mapping(rows[u'Modules_and_forms'])
+        return rows
```
this might be tricky. Setting headers and that mapping relies to have the `rows` data populated. so its like a series of steps that happen. Let me try to improve this while i move this out to a `PoFilesUtility`.

mkangia commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -0,0 +1,200 @@
+        self.translations = OrderedDict()
+        self.headers = dict()  # headers for each sheet name
+        self.sheet_name_to_module_or_form_type_and_id = dict()
+        self.generated_files = list()
```
what do you mean by a single resource here?

esoergel replied:
I mean one sheet/file - one set of translations, probably corresponding to a single form or module.

Right now, `translations`, `headers`, `sheet_name_to_module_or_form_type_and_id`, and `generated_files` each store a collection of data broken up by file.  For example, `self.translations` is a dict mapping file names to the translations for that file.

Instead, you could have a single property called `self.resources` (or whatever), which is a list of `Resource` objects (or a map of filenames to `Resource` objects).  Then each `Resource` object would manage it's own translations, headers, module/form type and id, and so forth.

This way, the broad strokes would happen in this parent class, and the child `Resource` class instances would handle the detail work, but they'd have a smaller set of state - only that corresponding to a single form or module.

mkangia replied:
hmm, all the translations and headers are received at once and `sheet_name_to_module_or_form_type_and_id` would also be generated at once from the first sheet. Its then essentially this method `_get_translation_for_sheet` that generates translation for a resource as needed for a po file, so it would be essentially be just this method in `Resource`. Will try that out to see if that makes sense to extract out.

mkangia commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -0,0 +1,200 @@
+            )
+
+    def get_filename(self, sheet_name):
+        return sheet_name + '_v' + str(self.version)
```
Replied to a similar thing above https://github.com/dimagi/commcare-hq/pull/20492#discussion_r187567030
good point about conflicts. Can add a timestamp to it to avoid conflict and also a quick view of when the file was created, though that timestamp exists in the file itself as well.
Ya not needed across requests.

mkangia commented on custom/icds/tasks.py:
```diff
@@ -44,3 +45,11 @@ def _get_query(db_name, max_age=max_age):
+    Transifex(domain, form_data.get('app_id'),
+              form_data.get('source_lang'),
+              form_data.get('target_lang'),
+              form_data.get('version')).send_files_to_transifex()
```
`send_files_to_transifex` is the single method call. It does all steps internally then.

mkangia commented on custom/icds_reports/forms.py:
```diff
@@ -0,0 +1,45 @@
+                                    choices=[('en', 'English')])
+    target_lang = forms.ChoiceField(label=ugettext_noop("Source Language"),
+                                    choices=[('hin', 'Hindi'), ('mr', 'Marathi'),
+                                             ('te', 'telugu')])
```
Good question. Have not thought of that but if it is to be done eventually, would not be app config but whats defined on transifex as the supported source languages. For now just added options expected for icds.

mkangia commented on corehq/apps/app_manager/dbaccessors.py:
```diff
@@ -431,11 +431,10 @@ def get_all_built_app_results(domain, app_id=None):
+    result = Application.get_db().view('app_manager/saved_app',
+                                       startkey=[domain, app_id, {}],
+                                       endkey=[domain, app_id],
+                                       descending=True)
```
looks like thats happening already, don't see docs included till i pass `include_docs=True`. Not sure why though

esoergel replied:
oh yeah, I think couchdbkit figures out which to use as the default for `include_docs` and `reduce` based on some set of parameters.  I can never remember which one is which, so I always specify if there's any confusion.

mkangia committed:
review fixes 3

- ensure available version in form
- redirect instead of render in case of success
- use get_current_app instead of get_app
- fix order of memoized and property

mkangia committed:
remove unused import [skip ci]

esoergel reviewed (approved):
thanks for the follow-up, looks good!

mkangia committed:
extract POFileGenerator

mkangia committed:
Merge branch 'master' into mk-icds-app-trans-hq

mkangia committed:
use constant

mkangia commented:
Thanks for the close reviews @esoergel . Was working on one more fix from last review, extracted POFilegenerator.

mkangia committed:
Merge master: resolve conflict in requirements.txt

mkangia committed:
fetch more details by add ?details to the url

mkangia committed:
add ability to upload translations as well.

mkangia committed:
fix check and add mapping for oriya

mkangia committed:
use tempfile instead of source code dir to store files

mkangia commented:
@esoergel let me know if you have more inputs on this, if g2g going to merge this now.

mkangia committed:
lint

mkangia commented on settings.py:
```diff
@@ -925,7 +925,7 @@
 
 MAX_RULE_UPDATES_IN_ONE_RUN = 10000
-
+TRANSIFEX_DETAILS = None
```
Added a todo for this

mkangia commented on custom/icds/translations/integrations/client.py:
```diff
@@ -0,0 +1,43 @@
+        return (self.username, self.token)
+
+    def list_resources(self):
+        url = "http://api.transifex.com/organizations/{}/projects/{}/resources".format(
```
tried that but looks like thats not supported. Got these links from their docs.

mkangia committed:
add todo [skip ci]

mkangia committed:
Improvement

- Cover all translations by ignoring duplicates
- update occurrences to be able to recreate xl file from translated po files

mkangia commented on corehq/apps/app_manager/app_translations/generators.py:
```diff
@@ -0,0 +1,175 @@
+            sheet_name_index = self._get_header_index(MODULES_AND_FORMS_SHEET_NAME, 'sheet_name')
+            unique_id_index = self._get_header_index(MODULES_AND_FORMS_SHEET_NAME, 'unique_id')
+
+            def occurrence(_row):
```
Not sure what to do here. Open to suggestions.

mkangia committed:
remove unused import

mkangia committed:
merge master: resolve conflict in toggles.py

mkangia committed:
update structure for teams to be under project

esoergel commented on corehq/apps/app_manager/app_translations/generators.py:
```diff
@@ -0,0 +1,174 @@
+        self.version = version
+        self.headers = dict()  # headers for each sheet name
+        self.generated_files = list()  # list of tuples (filename, filepath)
+        self.sheet_name_to_module_or_form_type_and_id = dict()
```
I do find all this business around sheets, rows, and column indices pretty confusing.  It's a little hard to follow, particularly since the application and transifex don't have anything to do with spreadsheets - that's just how we happen to have easy access to it, since that already exists.

You can punt on this if you don't think it's worth the effort, but you could modify `expected_bulk_app_sheet_rows` to return a custom data model that's easier to work with.

mkangia replied:
Agree.. but going to skip this for now and try to sneak this in during improvements later.

mkangia committed:
Add ori language [skip ci]

mkangia committed:
add msgctxt for duplicate source strings [skip ci]

mkangia commented on custom/icds/tasks.py:
```diff
@@ -49,8 +49,16 @@ def _get_query(db_name, max_age=max_age):
+                  data.get('target_lang'),
+                  data.get('transifex_project_slug'),
+                  data.get('version'),
+                  is_source_file=False).send_translation_files()
```
Yup, fixed later.

mkangia committed:
define as None instead of method def [skip ci]

mkangia commented on custom/icds/translations/integrations/client.py:
```diff
@@ -44,6 +44,22 @@ def upload_resource(self, path_to_pofile, resource_slug, resource_name):
+            'i18n_type': 'PO'
+        }
+        return requests.put(
+            url, data=json.dumps(data), auth=self._auth, headers=headers,
```
looks like that causing some issue when sending the request, possibly due to file content. So leaving this as-is.

mkangia committed:
add row index to msgctxt to differentiate duplicate msgctxt

mkangia committed:
limit length on case property for long calculations

mkangia commented on corehq/apps/app_manager/app_translations/generators.py:
```diff
@@ -96,7 +96,10 @@ def occurrence(_row):
+                    case_property = _row[case_property_index]
+                    # limit case property length to avoid errors at Transifex where there is a limit of 1000
+                    case_property = case_property[:950]
+                    return ':'.join([case_property, _row[list_or_detail_index]])
```
Can you share an example @esoergel . Sorry i didn't understand that. Are you referring to the length limit i added? If yes, that is to address very long case property calculations, i don't think anything else would be that long enough.

esoergel replied:
Yeah, I was thinking of the case_property truncation.  What do you mean "case property calculations"?  I was just thinking if there _are_ user-defined property names that are 1000 characters long, it's conceivable that the first 950 characters of two properties might be the same.  

Users frequently use patterns like this in their property names:
`patient_2_demographic_information_age`
`patient_2_demographic_information_sex`
But at any rate, it does seem pretty unlikely that this would happen at that length.

mkangia replied:
we found this long if expressions in excel sheet that was more than 1000 characters. Didn't come across a property name that was that long.. definitely highly unlikely someone would keep a name that long.. also looks like we wont rely on that to be unique. Might get more insight when i begin working on the second phase.

mkangia commented:
Going to merge this now... if there is still something pending can take it up in a following PR.
